src/parser/mod.rs

```diff
@@ -10408,13 +10408,86 @@ impl<'a> Parser<'a> {
             None
         };
 
+        let with_fill = if dialect_of!(self is ClickHouseDialect | GenericDialect)
+            && self.parse_keywords(&[Keyword::WITH, Keyword::FILL])
+        {
+            Some(self.parse_with_fill()?)
+        } else {
+            None
+        };
+
+        let interpolate = if dialect_of!(self is ClickHouseDialect | GenericDialect)
+            && self.parse_keyword(Keyword::INTERPOLATE)
+        {
+            if self.consume_token(&Token::LParen) {
+                if self.peek_token().token == Token::RParen {
+                    // INTERPOLATE ()
+                    self.next_token();
+                    Some(InterpolationArg::EmptyBody)
+                } else {
+                    // INTERPOLATE ( ... )
+                    let interpolations = self.parse_interpolations()?;
+                    self.expect_token(&Token::RParen)?;
+                    Some(InterpolationArg::Columns(interpolations))
+                }
+            } else {
+                // INTERPOLATE
+                Some(InterpolationArg::NoBody)
+            }
+        } else {
+            None
+        };
+
         Ok(OrderByExpr {
             expr,
             asc,
             nulls_first,
+            with_fill,
+            interpolate,
         })
     }
 
+    // Parse a WITH FILL clause (ClickHouse dialect)
+    // that follow the WITH FILL keywords in a ORDER BY clause
+    pub fn parse_with_fill(&mut self) -> Result<WithFill, ParserError> {
+        let from = if self.parse_keyword(Keyword::FROM) {
+            Some(self.parse_expr()?)
+        } else {
+            None
+        };
+
+        let to = if self.parse_keyword(Keyword::TO) {
+            Some(self.parse_expr()?)
+        } else {
+            None
+        };
+
+        let step = if self.parse_keyword(Keyword::STEP) {
+            Some(self.parse_expr()?)
+        } else {
+            None
+        };
+
+        Ok(WithFill { from, to, step })
+    }
+
+    // Parse a set of comma seperated INTERPOLATE expressions (ClickHouse dialect)
+    // that follow the INTERPOLATE keyword in an ORDER BY clause with the WITH FILL modifier
+    pub fn parse_interpolations(&mut self) -> Result<Vec<Interpolation>, ParserError> {
+        self.parse_comma_separated(|p| p.parse_interpolation())
+    }
+
+    // Parse a INTERPOLATE expression (ClickHouse dialect)
+    pub fn parse_interpolation(&mut self) -> Result<Interpolation, ParserError> {
+        let column = self.parse_expr()?;
+        let formula = if self.parse_keyword(Keyword::AS) {
+            Some(self.parse_expr()?)
+        } else {
+            None
+        };
```

Comment:
by reusing ExprWithAlias we'd likely be able to reuse the parsing function for that as well here https://github.com/sqlparser-rs/sqlparser-rs/blob/7022e4227fe00ae4914688ad91e355effd8a159b/src/parser/mod.rs#L9527-L9536

Reply:
I don't think we can use `ExprWithAlias` based on https://github.com/sqlparser-rs/sqlparser-rs/pull/1330#discussion_r1667012551